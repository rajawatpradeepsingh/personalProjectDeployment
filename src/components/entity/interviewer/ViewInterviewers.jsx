import { useState, useEffect, useCallback, useMemo, useRef } from "react";
import { useHistory } from "react-router-dom";
import { formatPhoneNumberIntl } from "react-phone-number-input";
import auth from "../../../utils/AuthService";
import AuthService from "../../../utils/AuthService";
import { getInterviewers, archiveSelectedInterviewers, getInterviewerByParams } from "../../../API/interviewers/interviewer-apis";
import { getAllClients } from "../../../API/clients/clients-apis";
import { getDict } from "../../../API/dictionaries/dictionary-apis";
import ReportModal from "../../../../src/components/utils/reports/report.modal";
import { reportTableFields } from "../../entity/interviewer/intvwr_utils/interviewerObjects";
import reportFormats from "../../utils/reports/reportUtils";
import Button from "../../common/button/button.component";
import { PageContainer } from "../../container/page-container/PageContainer";
import { PageHeader } from "../../container/page-header/PageHeader";
import { TableComponent } from "../../common/table/Table";
import { TableLink } from "../../common/table/TableLink";
import { RecordWithDetails } from "../../common/table/record_with_details/RecordWithDetails";
import { IntrvwrArchive } from "./intvwr_ui_helpers/archive_modal/IntrvwrArchive";
import { message } from "antd";
import { PlusOutlined } from "@ant-design/icons";
import IdleTimeOutHandler from "../../ui/Dashboard/IdleTimeOutHandler";




const ViewInterviewers = () => {
  const history = useHistory();
  const headers = useMemo(() => AuthService.getHeaders(), []);
  const archiveRef = useRef(null);
  const [currentPage, setCurrentPage] = useState(1);
  const [openArchive, setOpenArchive] = useState(false);
  const [showReportModal, setShowReportModal] = useState(false);
  const [interviewerList, setInterviewerList] = useState([]);
  const [totalItems, setTotalItems] = useState(0);
  const [pageSize, setPageSize] = useState(10);
  const [isLoading, setIsLoading] = useState(true);
  const [columns, setColumns] = useState([]);
  const [selectedRowKeys, setSelectedRowKeys] = useState([]);
  const [sort, setSort] = useState({});
  const [filters, setFilters] = useState({});
  const [searchOptions, setSearchOptions] = useState([]);
  const [clientFilters, setClientFilters] = useState([]);
  const [skillsFilter, setSkillsFilter] = useState([]);
  const [isActive, setIsActive] = useState(true);
  const [isLogout, setLogout] = useState(false);
  const user = useMemo(() => auth.getUserInfo(), []);
  const [roleData,setRoleData]=useState(user.roles[0]);
  const [isAdmin, setIsAdmin] = useState(false);
  const[managerRole,setManagerRole]= useState(false);
  console.log(roleData)
  useEffect(() => {
    if (auth.hasAdminRole()) setIsAdmin(true);
  }, []);
  const openEditPage = useCallback((id) => {
    history.push(`/interviewer/${id}`);
  }, [history]);

  const getClientFilters = useCallback(async () => {
    const res = await getAllClients(headers);
    if (res.statusCode === 200) {
      setClientFilters(res.tableData.map(client => ({
        text: `${client.clientName} (${client.address?.city})`,
        value: client.id
      })));
    } else {
      message.warning('Problem getting client filter data, refresh page or contact system admin')
    }
  }, [headers]);

  const getSkillOptions = useCallback(async () => {
    const response = await getDict("primeSkills", headers);
    if (response.length) {
      const primes = response.map((skill) => skill.value).sort();
      setSkillsFilter(primes.map((skill) => ({ text: skill, value: skill })));
    }
  }, [headers]);

  useEffect(() => getClientFilters(), [getClientFilters]);
  useEffect(() => getSkillOptions(), [getSkillOptions]);

  const defaultColumns = useMemo(
    () => [
      {
        title: "Name",
        dataIndex: "firstName",
        key: "firstName",
        sorter: true,
        fixed: "left",
        width: 170,
        render: (int, row) => (
          <TableLink
            onClick={() => openEditPage(row.id)}
            label={`${int} ${row.lastName}`}
          />
        ),
      },
      {
        title: "Contact",
        dataIndex: "email",
        key: "email",
        width: 120,
        render: (email, row) => (
          <RecordWithDetails
            name={email}
            location={row.phone_no ? formatPhoneNumberIntl(row.phone_no) : ""}
          />
        ),
      },
      {
        title: "Location",
        dataIndex: "address",
        key: "address.city",
        sorter: true,
        ellipsis: true,
        render: (add) => (
          <RecordWithDetails
            name={add.city || ""}
            location={`${add.state && `${add.state}, `}${add.country && `${add.country}`
              }`}
          />
        ),
        width: 180,
      },
      {
        title: "Skills",
        dataIndex: "interview_skills",
        key: "interview_skills",
        sorter: true,
        ellipsis: true,
        width: 170,
        filters: skillsFilter,
        filterSearch: true,
        render: (skills) => skills.split(",").join(", ")
      },
      {
        title: "Experience",
        dataIndex: "total_experience",
        key: "total_experience",
        render: (exp) => (
          <div style={{ textAlign: 'right', paddingRight: '10px' }}>{exp ? `${exp} year(s)` : ""}</div>
        ),
        width: 110,
        sorter: true,
      },
    
      {
        title: "Client",
        dataIndex: "client",
        key: "client.clientName",
        width: 150,
        sorter: true,
        render: (client) => (
          <RecordWithDetails
            name={client.clientName}
            location={client.address?.city}
          />
        ),
        filters: clientFilters,
        filterSearch: true,
      },
    ],
    [openEditPage, clientFilters, skillsFilter]
  );

  useEffect(() => {
    setColumns(defaultColumns);
  }, [defaultColumns, clientFilters, skillsFilter]);

  const loadInterviewers = useCallback(async () => {
  const pages = `pageNo=${currentPage}&pageSize=-1`;
    if (!Object.keys(filters).length) {
      let params = `?${pages}`;
      for (const key in sort) {
        if (key) {
          params += `&orderBy=${key}&orderMode=${sort[key]}`;
        }
      }
      const res = await getInterviewerByParams(headers, params);
      if (res) setIsLoading(false);
      if (res.statusCode === 200) {
        setInterviewerList(res.tableData);
        setTotalItems(res.totalItems);
      }
      if (res.statusCode !== 200) message.error("Something's gone wrong, refresh page or contact your system admin");
    } else {
      let sortParams = "";
      for (const key in sort) {
        if (key) {
          sortParams += `&orderBy=${key}&orderMode=${sort[key]}`;
        }
      }
      const filterBy = !Object.keys(filters).length ? null : filters;
      setIsLoading(true);
      const res = await getInterviewers(headers, currentPage, pageSize, sortParams ? sortParams : `&orderBy=id&orderMode=desc`  , filterBy);
      if (res) setIsLoading(false);
      if (res.status === 200) {
        setInterviewerList(res.data);
        setTotalItems(res.totalItems);
      }
      if (res.status !== 200) message.error("Something's gone wrong, refresh page or contact your system admin");
  }
  
  }, [headers, currentPage, pageSize, sort, filters]);

  useEffect(() => {
    loadInterviewers();
  }, [loadInterviewers, filters, archiveRef?.current?.length]);

  const archiveInterviewer = async () => {
    const res = await archiveSelectedInterviewers(headers, selectedRowKeys);
    if (!res.status.filter(s => s !== 200).length) {
      loadInterviewers();
      setSelectedRowKeys([]);
      archiveRef.current.loadArchive();
      message.success("Interviewer record archived successfully!");
    } else {
      message.error("Something's gone wrong, refresh page or contact your system admin");
    }
  };

  useEffect(() => {
    loadInterviewers();
  }, [loadInterviewers]);

  const searchInterviewer = useCallback((value) => setFilters({ ...filters, search: value }), [filters]);

  const getSearchSuggestions = useCallback((obj) => {
    let suggestions = [];
    for (let key in obj) {
      if (
        (typeof obj[key] === "string" || typeof obj[key] === "number") &&
        key !== "id" &&
        obj[key] !== ""
      ) {
        if (key === "interview_skills") {
          let values = obj[key].split(", ");
          suggestions.push(values[0], values[1], values[2]);
        } else {
          suggestions.push(obj[key]);
        }
      } else if (typeof obj[key] === "object") {
        suggestions = suggestions.concat(getSearchSuggestions(obj[key]));
      }
    }
    return suggestions;
  }, []);

  useEffect(() => {
    if (interviewerList) {
      let options = [];
      for (let obj of interviewerList) {
        let results = getSearchSuggestions(obj);
        results.forEach((item) => {
          if (!options.includes(item)) options.push(item);
        });
      }
      setSearchOptions(options);
    }
  }, [interviewerList, filters, getSearchSuggestions]);

  const openForm = () => {
    history.push("/addinterviewer");
  };

  const rowSelection = {
    onChange: (rowKeys, selectedRows) => {
      setSelectedRowKeys(rowKeys);
    },
    selectedRowKeys,
  };

  const cancelArchiving = () => {
    setSelectedRowKeys([]);
  };


  return (
    <PageContainer>
      <IdleTimeOutHandler
        onActive={() => { setIsActive(true) }}
        onIdle={() => { setIsActive(false) }}
        onLogout={() => { setLogout(true) }}
      />
      <PageHeader
        title="Interviewers"
        actions={
          (isAdmin || roleData.interviewsPermission ?
          <Button type="button" className="btn main" handleClick={openForm}>
            <PlusOutlined className="icon" />
            Add Interviewer
          </Button>:""
   ) }
      />
      <TableComponent
        loading={isLoading}
        data={interviewerList.map((int) => ({ ...int, key: int.id }))}
        columns={columns}
        setColumns={setColumns}
        defaultColumns={defaultColumns}
        rowSelection={isAdmin || roleData.interviewsPermission ? rowSelection :false}
        total={totalItems}
        setCurrentPage={setCurrentPage}
        setPageSize={setPageSize}
        setSort={setSort}
        setFilters={setFilters}
        filters={filters}
        searchOptions={searchOptions}
        searchList={interviewerList}
        handleSearch={searchInterviewer}
        openArchive={() => setOpenArchive(true)}
        openReports={() => setShowReportModal(true)}
        handleConfirmArchive={archiveInterviewer}
        handleCancelArchiving={cancelArchiving}
        removeableColumns={defaultColumns
          .filter((col) => col.title !== "Name" && col.title)
          .map((col) => col.title)}
      />

      <ReportModal
        showReportModal={showReportModal}
        setShowReportModal={setShowReportModal}
        individualReportName="interviewer"
        individualList={interviewerList}
        list={interviewerList.map((int) => ({
          id: int.id,
          name: `${int.firstName} ${int.lastName}`,
        }))}
        listLabel="Interviewers"
        filename="interviewer_report"
        tableFields={reportTableFields()}
        pdfFormatter={reportFormats.interviewerPdfFormat}
        csvExcelFormatter={reportFormats.interviewerCSVExcel}
      />

      <IntrvwrArchive
        open={openArchive}
        setOpen={setOpenArchive}
        ref={archiveRef}
      />
    </PageContainer>
  );
};

export default ViewInterviewers;