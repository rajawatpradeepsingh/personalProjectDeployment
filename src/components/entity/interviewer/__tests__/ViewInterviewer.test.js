import React from "react";
import axios from "axios";
import { ViewInterviewers } from "../ViewInterviewers";
import { render, cleanup, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

jest.mock("axios");
const mockShowModal = jest.fn();
const mockSetShowModal = jest.fn();
const mockEditInterviewer = jest.fn();

afterEach(cleanup);
afterEach(() => jest.resetAllMocks());
beforeEach(() => {
  sessionStorage.clear();
});

const adminInfo = {
  name: "user",
  pass: "newpass",
  firstName: "Test",
  lastName: "User",
  email: "xample@example.com",
  roles: [{ id: 1, roleName: "ADMIN" }],
};

const userNotAuth = {
  name: "user",
  pass: "newpass",
  firstName: "Test",
  lastName: "User",
  email: "xample@example.com",
  roles: [{ id: 1, roleName: "HR" }],
};

const data = {
  data: [
    {
      id: 1,
      full_name: "John Doe",
      location: "USA",
      phone_no: "+1(000) 000-0000",
      email: "test@test.com",
      interview_skills: "testing",
      current_project: "gap",
      total_experience: 3,
    },
    {
      id: 2,
      full_name: "Jane Doe",
      location: "OAK",
      phone_no: "+1(000) 222-2222",
      email: "test2@test.com",
      interview_skills: "technical",
      current_project: "rcp",
      total_experience: 10,
    },
  ],
};

describe("axios.get interviewers is successful", () => {
  it("renders component", async () => {
    let rendered;
    sessionStorage.setItem("userInfo", JSON.stringify(adminInfo));

    await act(async () => {
      await axios.get.mockResolvedValueOnce({
        URL: "http://localhost:9093/interviewer?pageNo=0&pageSize=10",
      });
      await axios.get.mockResolvedValueOnce(data);
      rendered = render(<ViewInterviewers />);
    });

    const heading = rendered.getByRole("heading");
    const table = rendered.getByRole("table");
    const thead = rendered.getAllByRole("columnheader");
    const trow = rendered.getAllByRole("row");
    const btns = rendered.getAllByRole("button");
    const editBtn = rendered.getAllByTestId("edit-btn");
    const checkboxes = rendered.getAllByRole("checkbox");

    expect(heading).toHaveTextContent("Interviewers");
    expect(table).toHaveProperty("className", "interviewerTable");
    expect(thead).toHaveLength(9);
    expect(trow).toHaveLength(3);
    expect(btns).toHaveLength(3);
    expect(editBtn).toHaveLength(2);
    expect(checkboxes).toHaveLength(2);
    expect(checkboxes[0]).not.toBeChecked();
    expect(checkboxes[1]).not.toBeChecked();
    expect(btns[0]).toBeDisabled();
  });

  it("renders comp with edit/delete btns disabled for non admin/recruiters", async () => {
    let rendered;
    sessionStorage.setItem("userInfo", JSON.stringify(userNotAuth));
    await act(async () => {
      await axios.get.mockResolvedValueOnce({
        URL: "http://localhost:9093/interviewer?pageNo=0&pageSize=10",
      });
      await axios.get.mockResolvedValueOnce(data);
      rendered = render(<ViewInterviewers />);
    });

    const editBtn = rendered.getAllByTestId("edit-btn");
    const checkboxes = rendered.getAllByRole("checkbox");

    expect(editBtn).toBeDisabled;
    expect(checkboxes).toBeDisabled;
  });

  it("matches snapshot DOM node structure", async () => {
    let rendered;
    sessionStorage.setItem("userInfo", JSON.stringify(adminInfo));

    await act(async () => {
      await axios.get.mockResolvedValueOnce({
        URL: "http://localhost:9093/interviewer?pageNo=0&pageSize=10",
      });
      await axios.get.mockResolvedValueOnce(data);
      rendered = render(<ViewInterviewers />);
    });

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("has correct table headers", async () => {
    let rendered;
    sessionStorage.setItem("userInfo", JSON.stringify(adminInfo));

    await act(async () => {
      await axios.get.mockResolvedValueOnce({
        URL: "http://localhost:9093/interviewer?pageNo=0&pageSize=10",
      });
      await axios.get.mockResolvedValueOnce(data);
      rendered = render(<ViewInterviewers />);
    });

    const thead = rendered.getAllByRole("columnheader");
    const trashIcon = rendered.getByTestId("trash-icon");

    expect(thead[0]).toContainElement(trashIcon);
    expect(thead[1]).toHaveTextContent("Edit");
    expect(thead[2]).toHaveTextContent("Name");
    expect(thead[3]).toHaveTextContent("Email");
    expect(thead[4]).toHaveTextContent("Contact #");
    expect(thead[5]).toHaveTextContent("Location");
    expect(thead[6]).toHaveTextContent("Interview Skills");
    expect(thead[7]).toHaveTextContent("Experience");
    expect(thead[8]).toHaveTextContent("Current Project");
  });

  it("data populates in correct fields", async () => {
    let rendered;
    sessionStorage.setItem("userInfo", JSON.stringify(adminInfo));

    await act(async () => {
      await axios.get.mockResolvedValueOnce({
        URL: "http://localhost:9093/interviewer?pageNo=0&pageSize=10",
      });
      await axios.get.mockResolvedValueOnce(data);
      rendered = render(<ViewInterviewers />);
    });

    const trow = rendered.getAllByRole("row");
    const cells = rendered.getAllByRole("cell");

    expect(trow[1]).toHaveTextContent("John Doe");
    expect(trow[1]).toHaveTextContent("+1(000) 000-0000");
    expect(trow[1]).toHaveTextContent("test@test.com");
    expect(trow[1]).toHaveTextContent("testing");
    expect(trow[1]).toHaveTextContent("gap");
    expect(trow[1]).toHaveTextContent(3);

    expect(trow[2]).toHaveTextContent("Jane Doe");
    expect(trow[2]).toHaveTextContent("+1(000) 222-2222");
    expect(trow[2]).toHaveTextContent("test2@test.com");
    expect(trow[2]).toHaveTextContent("technical");
    expect(trow[2]).toHaveTextContent("rcp");
    expect(trow[2]).toHaveTextContent(10);
  });

  it("clicking checkbox enables delete button", async () => {
    let rendered;
    sessionStorage.setItem("userInfo", JSON.stringify(adminInfo));

    await act(async () => {
      await axios.get.mockResolvedValueOnce({
        URL: "http://localhost:9093/interviewer?pageNo=0&pageSize=10",
      });
      await axios.get.mockResolvedValueOnce(data);
      rendered = render(<ViewInterviewers />);
    });

    const deleteBtn = rendered.getByTestId("int-delete");
    const checkboxes = rendered.getAllByRole("checkbox");

    userEvent.click(checkboxes[0]);

    expect(deleteBtn).toBeEnabled();
  });

  it("clicking edit button opens modal", async () => {
    let rendered;
    sessionStorage.setItem("userInfo", JSON.stringify(adminInfo));

    const selectedInterviewer = {
      id: 1,
      full_name: "John Doe",
      location: "USA",
      phone_no: "+1(000) 000-0000",
      email: "test@test.com",
      interview_skills: "testing",
      current_project: "gap",
      total_experience: 3,
    };

    await act(async () => {
      await axios.get.mockResolvedValueOnce({
        URL: "http://localhost:9093/interviewer?pageNo=0&pageSize=10",
      });
      await axios.get.mockResolvedValueOnce(data);
      rendered = render(<ViewInterviewers />);
    });

    const header = rendered.getAllByRole("heading");
    expect(header).toHaveLength(2);
    expect(header[1]).toHaveTextContent("Edit Interviewer");
  });
});

describe("axios.get interviewers fails", () => {
  it("loads empty table", async () => {
    let rendered;
    sessionStorage.setItem("userInfo", JSON.stringify(adminInfo));

    await act(async () => {
      await axios.get.mockResolvedValueOnce({
        URL: "http://localhost:9093/interviewer?pageNo=0&pageSize=10",
      });
      await axios.get.mockResolvedValueOnce([]);
      rendered = render(<ViewInterviewers />);
    });

    const thead = rendered.getAllByRole("columnheader");
    const trow = rendered.getAllByRole("row");
    const editBtn = screen.queryByRole("button", { name: "+" });
    const checkboxes = screen.queryByRole("checkbox");

    expect(thead).toHaveLength(9);
    expect(trow).toHaveLength(1);
    expect(editBtn).not.toBeInTheDocument();
    expect(checkboxes).not.toBeInTheDocument();
  });
});
