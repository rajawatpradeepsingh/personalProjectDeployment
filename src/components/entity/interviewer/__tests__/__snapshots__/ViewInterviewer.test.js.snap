// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`axios.get interviewers is successful matches snapshot DOM node structure 1`] = `
<DocumentFragment>
  <div
    class="comp-content"
  >
    <div
      class="comp-header"
    >
      <h1>
        Interviewers
      </h1>
    </div>
    <table
      class="interviewerTable"
    >
      <thead>
        <tr>
          <th
            class="muted"
          >
            <button
              class="list-delete"
              data-testid="int-delete"
              disabled=""
            >
              <svg
                class="trashcan"
                data-testid="trash-icon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M112 112l20 320c.95 18.49 14.4 32 32 32h184c17.67 0 30.87-13.51 32-32l20-320"
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="32"
                />
                <path
                  d="M80 112h352"
                  stroke-linecap="round"
                  stroke-miterlimit="10"
                  stroke-width="32"
                />
                <path
                  d="M192 112V72h0a23.93 23.93 0 0124-24h80a23.93 23.93 0 0124 24h0v40m-64 64v224m-72-224l8 224m136-224l-8 224"
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="32"
                />
              </svg>
            </button>
          </th>
          <th
            class="sr-only"
          >
            Edit
          </th>
          <th
            class="cell-text"
          >
            Name
          </th>
          <th
            class="cell-text"
          >
            Email
          </th>
          <th
            class="cell-int"
          >
            Contact #
          </th>
          <th
            class="cell-text"
          >
            Location
          </th>
          <th
            class="cell-text"
          >
            Interview Skills
          </th>
          <th
            class="cell-int"
          >
            Experience
          </th>
          <th
            class="cell-text"
          >
            Current Project
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="table-row"
        >
          <td
            class="btn-change"
          >
            <input
              class="check"
              type="checkbox"
            />
          </td>
          <td
            class="btn-change"
          >
            <button
              class="editBtn"
              data-testid="edit-btn"
              id="1"
              intid="1"
            />
          </td>
          <td
            class="cell-text"
          >
            John Doe
          </td>
          <td
            class="cell-text"
          >
            test@test.com
          </td>
          <td
            class="cell-int"
          >
            +1(000) 000-0000
          </td>
          <td
            class="int-location"
          >
             
          </td>
          <td
            class="cell-text"
          >
            testing
          </td>
          <td
            class="cell-int"
          >
            3 year(s)
          </td>
          <td
            class="cell-text"
          >
            gap
          </td>
        </tr>
        <tr
          class="table-row"
        >
          <td
            class="btn-change"
          >
            <input
              class="check"
              type="checkbox"
            />
          </td>
          <td
            class="btn-change"
          >
            <button
              class="editBtn"
              data-testid="edit-btn"
              id="2"
              intid="2"
            />
          </td>
          <td
            class="cell-text"
          >
            Jane Doe
          </td>
          <td
            class="cell-text"
          >
            test2@test.com
          </td>
          <td
            class="cell-int"
          >
            +1(000) 222-2222
          </td>
          <td
            class="int-location"
          >
             
          </td>
          <td
            class="cell-text"
          >
            technical
          </td>
          <td
            class="cell-int"
          >
            10 year(s)
          </td>
          <td
            class="cell-text"
          >
            rcp
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="pagination-row"
    />
  </div>
</DocumentFragment>
`;
